### PR TITLE
rewrite pausing to support enabling individual functions per wrapped token

### PR DESCRIPTION
**Pull request by botdad**

### CODE REVIEW

**Review by rweald — COMMENTED** [review]
How do you feel about maybe breaking the tests out into different files? It's getting decently hard to find things in an almost 1000 line long file. 

The new enable/disable functionality feels like a nice place to pull out a separate file just to contain those tests?

File: tests/canonical-swap.ts [hunk 1]
```diff
@@ -275,10 +276,111 @@ describe("canonical-swap", () => {
...
+          canonicalData: canonicalData.publicKey,
+          wrappedData: wrappedData.publicKey,
+          tokenProgram: TOKEN_PROGRAM_ID,
+        })
+        .signers([wallet.payer])
+        .rpc();
+
+      await expect(failedSwap).to.eventually.be.rejectedWith(
+        "AnchorError caused by account: wrapped_data. Error Code: ConstraintRaw. Error Number: 2003. Error Message: A raw constraint was violated."
+      );
```
Comment (jonorally) [inline tests/canonical-swap.ts]: Maybe worth adding an assert that swapWrappedForCanonical still succeeds here (and vice versa in `#swapWrappedForCanonicalEnabled` below)

**Review by jonorally — APPROVED** [review]

**Comment by rweald** [comment]
I might just be confused, but are we testing all potential paths in the enable/disable code? From my understanding there's: 

1. Authorized account + `enableWrappedToken(true)`
2. Authorized account + `enableWrappedToken(false)`
3. Not Authorized account + `enableWrappedToken(true)`
4. Not Authorized account + `enableWrappedToken(false)`

and then those 4 same paths / combos for the `disableWrappedToken` as well.  

Might be overkill, but given the risk and difficulty to publish changes feels worth exercising in tests all the possible paths. It's also possible I'm just following tests wrong and we do exercise all those paths.

**Comment by botdad** [comment]
Well since you’re both saying the same thing, I’ll add it.

There is an implicit test for the enabled state since we have successful swap tests, just not wrapped in the enable/disable calls.

**Comment by rweald** [comment]
I was actually a bit more worried about whether we explicitly test what happens if someone who isn't authorized attempts to call the `enable/disable` stuff.

File: programs/canonical-swap/src/lib.rs [hunk 2]
```diff
@@ -443,6 +444,7 @@ pub struct DisableWrappedToken<'info> {
 
     #[account(
         constraint = canonical_data.authority == *current_authority.key,
+        owner = *program_id,
```
Comment (jonorally) [inline programs/canonical-swap/src/lib.rs:447]: This is just a redundant check right? Isn't CanonicalData always going to be owned by the program (or I guess all this prevents is someone calling the program and messing up some unrelated fake canonical_data they passed in?)

**Review by jonorally — APPROVED** [review]

**Review by rweald — APPROVED** [review]
👏  much easier to follow the testing now.

File: programs/canonical-swap/src/lib.rs (hunk 2 above)
Comment (botdad) [reply programs/canonical-swap/src/lib.rs:447]: Yeah, I couldn't think of a way to get around this check, and i even think anchor does it for you automatically. There's no real perf loss by adding it and it's nice and clear so I did it.

Comment (jonorally) [reply programs/canonical-swap/src/lib.rs:447]: Yea makes sense, just wanted to make sure I didn't misunderstand some exploit haha

**Pull request merged**